subprojects/core/src/testFixtures/groovy/org/gradle/util/TestUtil.groovy

```diff
@@ -365,3 +366,68 @@ class TestUtil {
 interface TestClosure {
     Object call(Object param);
 }
+
+class TestProblems implements InternalProblems {
+    private final TestProblemSummarizer summarizer
+    private final InternalProblems delegate
+
+    TestProblems() {
+        this.summarizer = new TestProblemSummarizer()
+        this.delegate = new DefaultProblems(
+            summarizer,
+            null,
+            new TestCurrentBuildOperationRef(),
+            new ExceptionProblemRegistry(),
+            null
+        )
+    }
+
+    @Override
+    ProblemReporter getReporter() {
+        delegate.reporter
+    }
+
+    @Override
+    InternalProblemReporter getInternalReporter() {
+        delegate.internalReporter
+    }
+
+    @Override
+    AdditionalDataBuilderFactory getAdditionalDataBuilderFactory() {
+        delegate.additionalDataBuilderFactory
+    }
+
+    void assertProblemEmittedOnce(Object expectedProblem) {
+        assert summarizer.emitted.size() == 1
+        def problem = summarizer.emitted[0]
+        assert expectedProblem instanceof Closure ? expectedProblem.call(problem) : expectedProblem instanceof Wildcard ? true : expectedProblem == problem
```

Comment:
:x: this is really hard to read. Could you break this down into an if-else structure? I think doing that wouldn't break the `assert` statement.

Reply:
Fixed, please take another look.